### PR TITLE
Add route for topic identities and change cluster list format

### PR DESCRIPTION
**Pull request by kelseylam**
Added a route and method in KafkaStateCheck to get a list of topic identities for a given cluster. Wrote JSONW write methods for topic identities and related classes to convert them to jsons using toJSON. Changed clusters api call to not query based on status anymore and instead output a list of active clusters and a list of pending clusters.

### CODE REVIEW

File: app/controllers/api/KafkaStateCheck.scala [hunk 1]
```diff
@@ -42,31 +43,31 @@ class KafkaStateCheck (val messagesApi: MessagesApi, val kafkaManagerContext: Ka
...
-  def brokersSkewPercentage(c: String, t: String) = Action.async { implicit request =>
-    kafkaManager.getTopicIdentity(c, t).map { errorOrTopicIdentity =>
-      errorOrTopicIdentity.fold(
+  def clusters = Action.async { implicit request =>
+    implicit val formats = org.json4s.DefaultFormats
+    kafkaManager.getClusterList.map { errorOrClusterList =>
+      errorOrClusterList.fold(
         error => BadRequest(Json.obj("msg" -> error.msg)),
-        topicIdentity => Ok(Json.obj("topic" -> t, "brokersSkewPercentage" -> topicIdentity.brokersSkewPercentage))
+        clusterList => Ok(Serialization.writePretty("clusters" -> errorOrClusterList.getOrElse(None)))
```
Comment (jnewman) [inline app/controllers/api/KafkaStateCheck.scala]: Use `orNull` for this sort of thing instead of get or else

File: app/kafka/manager/features/KMFeature.scala [hunk 2]
```diff
@@ -75,4 +80,12 @@ object ClusterFeatures {
 
     ClusterFeatures(buffer.toSet)
   }
+
+  implicit val formats = Serialization.formats(FullTypeHints(List(classOf[ClusterFeatures])))
+
+  implicit def clusterFeaturesJSONW: JSONW[ClusterFeatures] = new JSONW[ClusterFeatures] {
+    def write(a: ClusterFeatures) =
+      makeObj(("features" -> toJSON(a.features.map(_.toString).toList))
```
Comment (jnewman) [inline app/kafka/manager/features/KMFeature.scala]: Why does this need three `toList`

File: app/controllers/api/KafkaStateCheck.scala [hunk 3]
```diff
@@ -42,31 +43,31 @@ class KafkaStateCheck (val messagesApi: MessagesApi, val kafkaManagerContext: Ka
...
-    status match {
-      case Some("active") => Json.obj("clusters" -> active.sortBy(_("name")))
-      case Some("pending") => Json.obj("clusters" -> pending.sortBy(_("name")))
-      case _ => Json.obj("clusters" -> (active ++ pending).sortBy(_("name")))
-    }
+    implicit val formats = org.json4s.DefaultFormats
+    Serialization.writePretty("topicIdentities" -> (for {
+      (tn, tiOpt) <- topicIdentities
+      ti <- tiOpt
+      } yield toJSON(tiOpt)))
```
Comment (jnewman) [inline app/controllers/api/KafkaStateCheck.scala]: Why not` toJSON(ti)`

File: app/controllers/api/KafkaStateCheck.scala [hunk 4]
```diff
@@ -7,11 +7,12 @@ package controllers.api
 
 import controllers.KafkaManagerContext
 import features.ApplicationFeatures
-import kafka.manager.model.ActorModel.KMClusterList
+import kafka.manager.model.ActorModel.{TopicIdentity}
```
Comment (realstraw) [inline app/controllers/api/KafkaStateCheck.scala]: nit: `{` and `}` is not required for single import

File: app/controllers/api/KafkaStateCheck.scala [hunk 5]
```diff
@@ -42,31 +43,31 @@ class KafkaStateCheck (val messagesApi: MessagesApi, val kafkaManagerContext: Ka
...
+        topicIdentityList => Ok(getTopicIdentitiesListJson(topicIdentityList.list))
       )
     }
   }
 
-  def getClusterListJson(clusterList: KMClusterList, status: Option[String]) = {
-    val active = clusterList.active.map(cc => Map("name" -> cc.name, "status" -> "active"))
-    val pending = clusterList.pending.map(cc => Map("name" -> cc.name, "status" -> "pending"))
+  def getTopicIdentitiesListJson(topicIdentities: IndexedSeq[(String, Option[TopicIdentity])]) = {
+    import org.json4s.scalaz.JsonScalaz._
```
Comment (realstraw) [inline app/controllers/api/KafkaStateCheck.scala]: which part of the following requires scalaz?

File: app/controllers/api/KafkaStateCheck.scala [hunk 6]
```diff
@@ -42,31 +43,31 @@ class KafkaStateCheck (val messagesApi: MessagesApi, val kafkaManagerContext: Ka
...
-        clusterList => Ok(getClusterListJson(clusterList, status))
+        topicIdentityList => Ok(getTopicIdentitiesListJson(topicIdentityList.list))
       )
     }
   }
 
-  def getClusterListJson(clusterList: KMClusterList, status: Option[String]) = {
-    val active = clusterList.active.map(cc => Map("name" -> cc.name, "status" -> "active"))
-    val pending = clusterList.pending.map(cc => Map("name" -> cc.name, "status" -> "pending"))
+  def getTopicIdentitiesListJson(topicIdentities: IndexedSeq[(String, Option[TopicIdentity])]) = {
```
Comment (jnewman) [inline app/controllers/api/KafkaStateCheck.scala]: I'd call this `topicIdentitiesAsJson` given the name above. The input type should just be Seq unless you need the special features of indexed seq.

File: app/controllers/api/KafkaStateCheck.scala (hunk 1 above)
Comment (realstraw) [inline app/controllers/api/KafkaStateCheck.scala]: does `errorOrClusterList.toOption` work? instead of `errorOrClusterList.getOrElse(None))`

File: app/controllers/api/KafkaStateCheck.scala (hunk 3 above)
Comment (jnewman) [inline app/controllers/api/KafkaStateCheck.scala]: This would be nicer like `topicIds.map(_._1).filter(_.nonEmpty)`

Comment (realstraw) [inline app/controllers/api/KafkaStateCheck.scala]: Inspired by @jnewman , if `ti` is what we after instead of `tiOpt`, we could also use `topicIds.map(_._1).flatten` ;-)

Also sure why we choose json4s here where the rest of this class seems to use play json. But if we do want to use Json4s, alternatively, this whole serialization part could be simplified to use the json4s DSL, e.g.

```scala
import org.json4s.JsonDSL._
import org.json4s.jackson.JsonMethods._

pretty(render("topicIdentities" -> topicIds.map(_._1).flatten))
```

File: app/kafka/manager/features/KMFeature.scala [hunk 7]
```diff
@@ -75,4 +80,12 @@ object ClusterFeatures {
 
     ClusterFeatures(buffer.toSet)
   }
+
+  implicit val formats = Serialization.formats(FullTypeHints(List(classOf[ClusterFeatures])))
+
+  implicit def clusterFeaturesJSONW: JSONW[ClusterFeatures] = new JSONW[ClusterFeatures] {
```
Comment (realstraw) [inline app/kafka/manager/features/KMFeature.scala]: why do we need this `JSONW` here?

File: app/controllers/api/KafkaStateCheck.scala [hunk 8]
```diff
@@ -42,31 +43,31 @@ class KafkaStateCheck (val messagesApi: MessagesApi, val kafkaManagerContext: Ka
     }
   }
 
-  def clusters(status: Option[String]) = Action.async { implicit request =>
-    kafkaManager.getClusterList.map { errorOrClusterList =>
-      errorOrClusterList.fold(
+  def topicIdentities(c: String) = Action.async { implicit request =>
```
Comment (jnewman) [inline app/controllers/api/KafkaStateCheck.scala]: `cluster` would make a better name, I think

File: app/kafka/manager/features/KMFeature.scala (hunk 7 above)
Comment (realstraw) [inline app/kafka/manager/features/KMFeature.scala]: I don't know too well about these JSONW conversions, and I might miss something, but it seems to me, that for most of the cases I see here, you could simple do the following after you have the right implicit format
```scala
import org.json4s.native.Serialization.write

val theCaseClass: ClusterFeatures = ???
val jsonObject = write(theCaseClass)
```

Since I saw the keys you are using for the json object are the case class field names anyway, the above implementation will use the case class field names so you don't have to map all the names to a string format manually as you did in all the `write` methods in these JSONW objects anymore.

File: app/kafka/manager/features/KMFeature.scala (hunk 2 above)
Comment (kelseylam) [inline app/kafka/manager/features/KMFeature.scala]: I don't think toJSON implicitly works on Sets.

File: app/controllers/api/KafkaStateCheck.scala (hunk 3 above)
Comment (kelseylam) [inline app/controllers/api/KafkaStateCheck.scala]: If we just use play json as it's been used before, like
` Ok(Json.obj("topicIdentities" -> topicIdentityList.list.map(_._2).flatten.toList))
`
it says it expects a (String, Json.JsValueWrapper) and it's getting back a (String, List[TopicIdentity]) instead. 

For some reason using pretty and render won't work either, which is why we used Serialization.writePretty. 

image

If you know why this problem occurred or any alternatives we would really appreciate it.

File: app/controllers/api/KafkaStateCheck.scala (hunk 8 above)
Comment (kelseylam) [inline app/controllers/api/KafkaStateCheck.scala]: Since in the routes and other KafkaStateCheck methods topics and clusters are referred to as "c" and "t", we wanted to be consistent with the existing code.

File: app/kafka/manager/model/ActorModel.scala [hunk 9]
```diff
@@ -409,9 +410,37 @@ import scala.language.reflectiveCalls
...
     import org.json4s.scalaz.JsonScalaz._
+    import org.json4s._
+    import org.json4s.jackson.Serialization
 
 import scala.language.reflectiveCalls
-    
+
+    implicit val formats = Serialization.formats(FullTypeHints(List(classOf[TopicIdentity])))
+
+    implicit def topicIdentityJSONW: JSONW[TopicIdentity] = new JSONW[TopicIdentity] {
```
Comment (realstraw) [inline app/kafka/manager/model/ActorModel.scala]: why is this necessary?

Comment (kelseylam) [inline app/kafka/manager/model/ActorModel.scala]: If you're asking about topicIdentityJSONW, it's so we can use toJSON to convert a TopicIdentity into a JValue that has attributes besides the ones in Topic Identity's case class constructor (eg replicationFactor, brokersSkewPercentage).

Comment (realstraw) [inline app/kafka/manager/model/ActorModel.scala]: I think
```
import org.json4s.native.Serialization.write
write(someTopicIdentity)
```
would produce the same thing without the `JSONW` implicit, which seems easier to me. It uses the case class's field name as the json object's keys, which is safer and easier than what you manually specify in the `JSONW` conversion.

File: app/controllers/api/KafkaStateCheck.scala [hunk 10]
```diff
@@ -42,31 +43,24 @@ class KafkaStateCheck (val messagesApi: MessagesApi, val kafkaManagerContext: Ka
...
+    implicit val formats = org.json4s.DefaultFormats
+    import org.json4s.scalaz.JsonScalaz._
+
+    kafkaManager.getTopicListExtended(c).map { errorOrTopicList =>
+      errorOrTopicList.fold(
         error => BadRequest(Json.obj("msg" -> error.msg)),
-        clusterList => Ok(getClusterListJson(clusterList, status))
+        topicIdentityList => Ok(Serialization.writePretty("topicIdentities" -> topicIdentityList.list.map(tup => toJSON(tup._2))))
```
Comment (kelseylam) [inline app/controllers/api/KafkaStateCheck.scala]: @jnewman @realstraw I was wondering if either of you had any opinions on how to best write this line. topicIdentityList.list is an IndexedSeq[String, Option[ActorModel.TopicIdentity]]. After the `"topicIdentities" ->`, I want to get back a list of defined TopicIdentity objects with toJSON called on each one of them. 

For some reason, after using map I can't use flatten.

Comment (realstraw) [inline app/controllers/api/KafkaStateCheck.scala]: you cannot use `flatten` is because it has to be an `Iterable` of `Iterable` to use flatten, after map it becomes an `Iterable` of `JValue`

Comment (jnewman) [inline app/controllers/api/KafkaStateCheck.scala]: Something like:

```scala
topicIdentityList.flatMap(_._2).map(toJson(_))
```

For example,
```scala
val iS = IndexedSeq[(String, Option[Int])](("foo", Option(42)), ("foo", None), ("foo", Option(99)))
iS.flatMap(_._2).map(_.toString)
// res0: IndexedSeq[String] = Vector(42, 99)
```

File: app/kafka/manager/features/KMFeature.scala [hunk 11]
```diff
@@ -53,6 +53,8 @@ object ClusterFeature extends Logging {
 case class ClusterFeatures(features: Set[ClusterFeature])
 
 object ClusterFeatures {
+  import scala.language.reflectiveCalls
```
Comment (jnewman) [inline app/kafka/manager/features/KMFeature.scala]: Do you need this?

File: app/kafka/manager/model/ActorModel.scala (hunk 9 above)
Comment (jnewman) [inline app/kafka/manager/model/ActorModel.scala]: I think the issue was that the case class didn't have all the properties in the case class constructor, which meant either decorating the case class or listing the props out. (The latter had been done elsewhere in this project; I'd initially suggested decorating the case class.)

File: app/controllers/api/KafkaStateCheck.scala [hunk 12]
```diff
@@ -46,11 +45,10 @@ class KafkaStateCheck (val messagesApi: MessagesApi, val kafkaManagerContext: Ka
...
     implicit val formats = org.json4s.DefaultFormats
     import org.json4s.scalaz.JsonScalaz._
-
-    kafkaManager.getTopicListExtended(c).map { errorOrTopicList =>
-      errorOrTopicList.fold(
+    kafkaManager.getTopicListExtended(c).map { errorOrTopicListExtended =>
+      errorOrTopicListExtended.fold(
         error => BadRequest(Json.obj("msg" -> error.msg)),
-        topicIdentityList => Ok(Serialization.writePretty("topicIdentities" -> topicIdentityList.list.map(tup => toJSON(tup._2))))
+        topicListExtended => Ok(Serialization.writePretty("topicIdentities" -> topicListExtended.list.map {case (_, topicIdentity) => toJSON(topicIdentity)}))
```
Comment (jnewman) [inline app/controllers/api/KafkaStateCheck.scala]: You're still calling toJson on an Option, right?

File: app/controllers/api/KafkaStateCheck.scala [hunk 13]
```diff
@@ -42,31 +42,23 @@ class KafkaStateCheck (val messagesApi: MessagesApi, val kafkaManagerContext: Ka
...
-
-  def brokersSkewPercentage(c: String, t: String) = Action.async { implicit request =>
-    kafkaManager.getTopicIdentity(c, t).map { errorOrTopicIdentity =>
-      errorOrTopicIdentity.fold(
+  def clusters = Action.async { implicit request =>
+    implicit val formats = org.json4s.DefaultFormats
+    kafkaManager.getClusterList.map { errorOrClusterList =>
+      errorOrClusterList.fold(
         error => BadRequest(Json.obj("msg" -> error.msg)),
-        topicIdentity => Ok(Json.obj("topic" -> t, "brokersSkewPercentage" -> topicIdentity.brokersSkewPercentage))
```
Comment (jnewman) [inline app/controllers/api/KafkaStateCheck.scala]: This was something we added previously?

File: app/controllers/api/KafkaStateCheck.scala (hunk 12 above)
Comment (kelseylam) [inline app/controllers/api/KafkaStateCheck.scala]: I haven't gotten nulls so far but that might be because we're testing on local data so there are probably no nulls. I just saw your comment about flatMap and it works so I can use that if you think that will take care of the potential null problem?

File: app/controllers/api/KafkaStateCheck.scala (hunk 13 above)
Comment (kelseylam) [inline app/controllers/api/KafkaStateCheck.scala]: Yes, it was just for basic testing but didn't end up being useful.

File: app/kafka/manager/model/ActorModel.scala (hunk 9 above)
Comment (realstraw) [inline app/kafka/manager/model/ActorModel.scala]: Add the `FieldSerializer[TopicIdentity]` to the implicit format will put all `val` defined in the case class to json fields. It might be a big change, so I'm good with the current implementation

Comment (kelseylam) [inline app/kafka/manager/model/ActorModel.scala]: FieldSerializer worked, but we actually also used toJSON to format one of the TopicIdentity's attributes (partitionsIdentity) differently. Since that attribute is a val and immutable, I don't think we could've changed it's formatting with any better way than using toJSON which is why we are still using toJSON.

Comment (realstraw) [inline app/kafka/manager/model/ActorModel.scala]: 👍

File: app/controllers/api/KafkaStateCheck.scala [hunk 14]
```diff
@@ -42,31 +42,23 @@ class KafkaStateCheck (val messagesApi: MessagesApi, val kafkaManagerContext: Ka
     }
   }
 
-  def clusters(status: Option[String]) = Action.async { implicit request =>
-    kafkaManager.getClusterList.map { errorOrClusterList =>
-      errorOrClusterList.fold(
+  def topicIdentities(c: String) = Action.async { implicit request =>
+    implicit val formats = org.json4s.DefaultFormats
+    import org.json4s.scalaz.JsonScalaz._
```
Comment (jnewman) [inline app/controllers/api/KafkaStateCheck.scala]: What's this needed for? (Just curious)

**Comment by jnewman** [comment]
LGTM side from the one import I don't understand.

File: app/controllers/api/KafkaStateCheck.scala (hunk 14 above)
Comment (kelseylam) [inline app/controllers/api/KafkaStateCheck.scala]: The toJSON method.

**Comment by realstraw** [comment]
👍

**Comment by arijitdey** [comment]
👍

**Comment by jnewman** [comment]
:+1:
